File: chapters/chapter_04_cpp_pubsub/examples/simple_publisher.cpp
#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

/**
 * Simple Publisher Node in C++
 * Publishes "Hello World" messages to /chatter topic
 */
class SimplePublisher : public rclcpp::Node
{
public:
    SimplePublisher() : Node("simple_publisher"), counter_(0)
    {
        // Create publisher
        publisher_ = this->create_publisher<std_msgs::msg::String>("chatter", 10);
        
        // Create timer (500ms)
        timer_ = this->create_wall_timer(
            500ms,
            std::bind(&SimplePublisher::timer_callback, this));
        
        RCLCPP_INFO(this->get_logger(), "Simple Publisher has started!");
    }

private:
    void timer_callback()
    {
        auto message = std_msgs::msg::String();
        message.data = "Hello World: " + std::to_string(counter_);
        
        RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
        publisher_->publish(message);
        
        counter_++;
    }
    
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    size_t counter_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SimplePublisher>());
    rclcpp::shutdown();
    return 0;
}

File: chapters/chapter_03_python_pubsub/examples/simple_publisher.py
#!/usr/bin/env python3
"""
Simple Publisher Node
Publishes "Hello World" messages to /chatter topic
"""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class SimplePublisher(Node):
    """
    A simple publisher node that sends messages periodically.
    
    This demonstrates:
    - Creating a ROS2 node class
    - Creating a publisher
    - Using a timer for periodic execution
    """
    
    def __init__(self):
        # Initialize the node with a name
        super().__init__('simple_publisher')
        
        # Create a publisher
        # - Message type: String
        # - Topic name: '/chatter'
        # - Queue size: 10 (how many messages to buffer)
        self.publisher_ = self.create_publisher(String, 'chatter', 10)
        
        # Create a timer that calls our callback every 0.5 seconds
        timer_period = 0.5  # seconds
        self.timer = self.create_timer(timer_period, self.timer_callback)
        
        # Counter for our messages
        self.counter = 0
        
        # Log that we've started
        self.get_logger().info('Simple Publisher has started!')
    
    def timer_callback(self):
        """
        This function is called every 0.5 seconds by the timer.
        It creates and publishes a message.
        """
        # Create a message
        msg = String()
        msg.data = f'Hello World: {self.counter}'
        
        # Publish the message
        self.publisher_.publish(msg)
        
        # Log what we published
        self.get_logger().info(f'Publishing: "{msg.data}"')
        
        # Increment counter
        self.counter += 1


def main(args=None):
    """Main function to start the node"""
    # Initialize the ROS2 Python library
    rclpy.init(args=args)
    
    # Create our publisher node
    node = SimplePublisher()
    
    # Keep the node running (and calling callbacks)
    # This blocks until you press Ctrl+C
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    
    # Cleanup
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
